modules/core/src/lbfgs.cpp

```diff
@@ -0,0 +1,427 @@
+//
+// This software is provided by the copyright holders and contributors "as is" and
+// any express or implied warranties, including, but not limited to, the implied
+// warranties of merchantability and fitness for a particular purpose are disclaimed.
+// In no event shall the OpenCV Foundation or contributors be liable for any direct,
+// indirect, incidental, special, exemplary, or consequential damages
+// (including, but not limited to, procurement of substitute goods or services;
+// loss of use, data, or profits; or business interruption) however caused
+// and on any theory of liability, whether in contract, strict liability,
+// or tort (including negligence or otherwise) arising in any way out of
+// the use of this software, even if advised of the possibility of such damage.
+//
+//M*/
+
+#include "precomp.hpp"
+
+namespace cv
+{
+    class LBFGSSolverImpl : public LBFGSSolver
+    {
+    public:
+        Ptr<Function> getFunction() const;
+        void setFunction(const Ptr<Function>& f);
+        TermCriteria getTermCriteria() const;
+        LBFGSSolverImpl();
+        void setTermCriteria(const TermCriteria& termcrit);
+        double minimize(InputOutputArray x);
+
+    private:
+        double linesearch(const Mat &x, const Mat &searchDir, double a);
+        int cvsrch(Mat &x, double f, Mat &g, double &stp, Mat &s);
+        static int cstep(double& stx, double& fx, double& dx, double& sty, double& fy, double& dy, double& stp,
+                         const double fp, const double dp, bool& brackt, const double stpmin, const double stpmax, int& info);
+
+    protected:
+        Ptr<MinProblemSolver::Function> _Function;
+        TermCriteria _termcrit;
+    };
+
+    LBFGSSolverImpl::LBFGSSolverImpl() {
+        _Function = Ptr<Function>();
+    }
+
+    Ptr<MinProblemSolver::Function> LBFGSSolverImpl::getFunction() const {
+        return _Function;
+    }
+
+    void LBFGSSolverImpl::setFunction(const Ptr<Function>& f) {
+        _Function = f;
+    }
+
+    TermCriteria LBFGSSolverImpl::getTermCriteria() const {
+        return _termcrit;
+    }
+
+    void LBFGSSolverImpl::setTermCriteria(const TermCriteria& termcrit) {
+        CV_Assert(
+            (termcrit.type == (TermCriteria::MAX_ITER + TermCriteria::EPS) && termcrit.epsilon > 0 && termcrit.maxCount > 0) ||
+            ((termcrit.type == TermCriteria::MAX_ITER) && termcrit.maxCount > 0)
+        );
+        _termcrit = termcrit;
+    }
+
+    double LBFGSSolverImpl::linesearch(const Mat &x, const Mat &searchDir, double a) {
+        double ak = a;
+        double fval = _Function->calc(x.ptr<double>());
+        Mat g;
+        x.copyTo(g);
+        _Function->getGradient(x.ptr<double>(), g.ptr<double>());
+
+        Mat s;
+        searchDir.copyTo(s);
+        Mat xx;
+        x.copyTo(xx);
+
+        cvsrch(xx, fval, g, ak, s);
+
+        return ak;
+    }
+
+    int LBFGSSolverImpl::cvsrch(Mat &x, double f, Mat &g, double &stp, Mat &s) {
```

Comment:
`LBFGSSolverImpl` share a lot of code with `LBFGSSolverImpl`. The only difference is `minimize()` function. So IMHO it would be better to inherit both from the same base class to simplify further support.